File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        if len(nums) == 1:
            return nums[0]        

        n = len(nums)
        return max(self.getMaxValue(nums, 0, n-1), self.getMaxValue(nums, 1, n))

    def getMaxValue(self, nums, start, end):

        prev1, prev2 = 0,0

        for i in range(start, end):
            temp = prev1
            prev1 = max(prev1, prev2 + nums[i])
            prev2 = temp
        
        return max(prev1, prev2)

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {

    public int rob(int[] nums) {

        if(nums.length == 1){
            return nums[0];
        }

        int n = nums.length;

        return Math.max(getMaxAmount(0, n-1, nums), getMaxAmount(1, n, nums));
    }

    private int getMaxAmount(int start, int end, int[] nums){
        int prev = 0;
        int curr = 0;
        for(int i = start; i < end; i++){
            int temp = curr;
            curr = Math.max(curr, prev + nums[i]);
            prev = temp;
        }

        return Math.max(prev, curr); 
    }    
}